hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/ConditionalWriteLockProvider.java

```diff
@@ -0,0 +1,581 @@
+      return SUPPORTED_LOCK_SERVICE_IMPLEMENTATIONS.get(scheme);
+    } else {
+      throw new HoodieNotSupportedException("No implementation of StorageLock supports this scheme: " + scheme);
+    }
+  }
+
+  @VisibleForTesting
+  ConditionalWriteLockProvider(
+      int heartbeatIntervalMs,
+      int lockValidityMs,
+      String bucketName,
+      String lockFilePath,
+      String ownerId,
+      HeartbeatManager heartbeatManager,
+      StorageLock lockService,
+      Logger logger) {
+    this.heartbeatIntervalMs = heartbeatIntervalMs;
+    this.lockValidityMs = lockValidityMs;
+    this.bucketName = bucketName;
+    this.lockFilePath = lockFilePath;
+    this.heartbeatManager = heartbeatManager;
+    this.lockService = lockService;
+    this.ownerId = ownerId;
+    this.logger = logger;
+    logger.debug("Instantiating new Conditional Write LP, owner: {}", ownerId);
+  }
+
+  // -----------------------------------------
+  // BASE METHODS
+  // -----------------------------------------
+
+  @Override
+  public synchronized StorageLockFile getLock() {
+    return currentLockObj;
+  }
+
+  /**
+   * Attempts to acquire the lock within the given timeout.
+   */
+  @Override
+  public boolean tryLock(long time, TimeUnit unit) {
+    long deadlineNanos = System.nanoTime() + unit.toNanos(time);
+
+    while (System.nanoTime() < deadlineNanos) {
+      try {
+        if (tryLock()) {
+          return true;
+        }
+        logger.debug(
+            LOCK_STATE_LOGGER_MSG,
+            ownerId,
+            lockFilePath,
+            Thread.currentThread(),
+            ACQUIRING);
+        Thread.sleep(DEFAULT_LOCK_ACQUISITION_BUFFER);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new HoodieLockException(generateLockStateMessage(LockState.FAILED_TO_ACQUIRE), e);
+      }
+    }
+
+    return false;
+  }
+
+  @Override
+  public synchronized void close() {
+    try {
+      this.unlock();
+    } catch (Exception e) {
+      logger.error("Owner {}: Failed to unlock current lock.", ownerId, e);
+    }
+    try {
+      this.lockService.close();
+    } catch (Exception e) {
+      logger.error("Owner {}: Lock service failed to close.", ownerId, e);
+    }
+    try {
+      this.heartbeatManager.close();
+    } catch (Exception e) {
+      logger.error("Owner {}: Heartbeat manager failed to close.", ownerId, e);
+    }
+
+    this.isClosed = true;
+  }
+
+  private synchronized boolean isLockStillValid(StorageLockFile lock) {
+    return !lock.isExpired() && !isCurrentTimeCertainlyOlderThanDistributedTime(lock.getValidUntil());
+  }
+
+  /**
+   * Attempts a single pass to acquire the lock (non-blocking).
+   * 
+   * @return true if lock acquired, false otherwise
+   */
+  @Override
+  public synchronized boolean tryLock() {
+    assertHeartBeatManagerExists();
+    assertUnclosed();
+    logger.debug(LOCK_STATE_LOGGER_MSG, ownerId, lockFilePath, Thread.currentThread(), ACQUIRING);
```

Comment:
Should the logging using `LOCK_STATE_LOGGER_MSG` and `LOCK_STATE_LOGGER_MSG_WITH_INFO` be wrapped into methods and only taking the lock state and extra info (`ownerId, lockFilePath, Thread.currentThread()` are pre-determined)?  Otherwise the string (`LOCK_STATE_LOGGER_MSG`) is opaque and not easy to use directly.

Reply:
I want to avoid a switch with some enum to decide whether to logger.debug, logger.info, etc.

I also need to avoid generating the string before calling logger.debug(xxx) as the string will be created at runtime regardless if I take this approach